### PR TITLE
[HIVEMALL-74] Implement pLSA

### PR DESCRIPTION
**Pull request by takuti**
## What changes were proposed in this pull request?

Implement (incremental) probabilistic latent semantic analysis (pLSA) algorithm:

- Original papers:
  - [Probabilistic Latent Semantic Indexing](http://dl.acm.org/citation.cfm?id=312649)
  - [Probabilistic Latent Semantic Analysis](http://www.iro.umontreal.ca/~nie/IFT6255/Hofmann-UAI99.pdf)
- Incremental variant which is implemented in this PR:
  - [Incremental Probabilistic Latent Semantic Analysis for Automatic Question Recommendation](https://pdfs.semanticscholar.org/b66e/c7faf2e4888503f7ad1537d284f350fb3e58.pdf)

## What type of PR is it?

Feature

## What is the Jira issue?

https://issues.apache.org/jira/browse/HIVEMALL-74

## How was this patch tested?

- unit tests
- manual tests on local

## How to use this feature?

Basically, interfaces are similar to Online LDA implemented in #66; we have `train_plsa()` and `plsa_predict()` functions.

For a `docs` table:

| docid | doc  |
|:---:|:---|
| 1  | "Fruits and vegetables are healthy." |
|2 | "I like apples, oranges, and avocados. I do not like the flu or colds." |

the following query successfully learns topics behind the two documents:

```sql
with word_counts as (
  select
    docid,
    feature(word, count(word)) as f
  from docs t1 lateral view explode(tokenize(doc, true)) t2 as word
  where
    not is_stopword(word)
  group by
    docid, word
),
plsa_model as (
  select
    train_plsa(feature, "-topic 2 -iter 10000 -eps 0.00001 -delta 0.00001 -alpha 0.00001") as (label, word, prob)
  from (
    select docid, collect_set(f) as feature
    from word_counts
    group by docid
  ) t
),
test as (
  select
    docid,
    word,
    count(word) as value
  from docs t1 LATERAL VIEW explode(tokenize(doc, true)) t2 as word
  where
    not is_stopword(word)
  group by
    docid, word
),
topic as (
  select
    t.docid,
    plsa_predict(t.word, t.value, m.label, m.prob, "-topic 2 -delta 0.01 -alpha 0.00001") as probabilities
  from
    test t
    JOIN plsa_model m ON (t.word = m.word)
  group by
    t.docid
)
select docid, probabilities, probabilities[0].label, m.words -- topic each document should be assigned
from topic t
join (
  select label, collect_set(feature(word, prob)) as words
  from plsa_model
  group by label
) m on t.probabilities[0].label = m.label
;
```

|docid  | probabilities  | label |  words |
|:---:|:---|:---:|:---|
|1   |    [{"label":0,"probability":1.0},{"label":1,"probability":1.1246405E-32}]  | 0|["fruits:0.33333066","healthy:0.33333066","vegetables:0.33333066","like:2.4555745E-6","avocados:2.0867487E-6","colds:1.665714E-6","flu:1.0358361E-6","apples:5.5809795E-7","oranges:2.2456922E-7"] |
|2 |      [{"label":1,"probability":0.9999961},{"label":0,"probability":3.886718E-6}]  |   1     |  ["like:0.28571412","oranges:0.14285718","colds:0.14285718","avocados:0.14285718","apples:0.14285718","flu:0.14285718","healthy:1.841767E-32","vegetables:1.2376679E-32","fruits:7.812756E-34"] |

We can observe that topics obtained from pLSA overfits to the training documents, compared to LDA.

Note that choosing a hyperparameter `alpha` might be difficult; inappropriate values lead undesirable P(w|z).

### CODE REVIEW

**Comment by takuti** [comment]
Since interfaces are similar to Online LDA, this PR should be handled after #66 is merged.

**Comment by myui** [comment]
@takuti please update this PR following the way in LDA.

File: core/src/main/java/hivemall/topicmodel/IncrementalPLSAModel.java [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    private void initParams() {
+        final List<float[]> p_dz = new ArrayList<float[]>();
+        final List<Map<String, float[]>> p_dwz = new ArrayList<Map<String, float[]>>();
+
+        for (int d = 0; d < _miniBatchSize; d++) {
+            // init P(z|d)
+            float[] p_dz_d = ArrayUtils.newRandomFloatArray(_K, _rnd);
+            ArrayUtils.normalize(p_dz_d);
```
Comment (myui) [inline core/src/main/java/hivemall/topicmodel/IncrementalPLSAModel.java]: `normalize` should be moved to `MathUtils`, not a normal array operation.

File: core/src/main/java/hivemall/topicmodel/IncrementalPLSAModel.java [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+            double npSumInDoc_zw_w = 0.d; // sum over the labels in the document
+            double pSumAll_zw_w = 0.d; // sum over the all existing labels
+
+            for (Map.Entry<String, float[]> e : _p_zw.entrySet()) {
+                final String label = e.getKey();
+                final float[] p_zw_w = e.getValue();
+
+                if (doc.containsKey(label)) {
+                    p_zw_w[z] *= _alpha; // alpha * P(w|z)
+                    float np = doc.get(label) * p_dwz_d.get(label)[z]; // n(d,w) * P(z|d,w)
```
Comment (myui) [inline core/src/main/java/hivemall/topicmodel/IncrementalPLSAModel.java]: Avoid autoboxing `doc.get(label)`.

```java
Float label_value = doc.get(label);
if(label_value == null) {
   ..
} else {
  p_zw_w[z] *= _alpha; // alpha * P(w|z)
  float np = label_value.floatValue() * p_dwz_d.get(label)[z];
```

File: core/src/main/java/hivemall/topicmodel/IncrementalPLSAModel.java [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+            ArrayUtils.normalize(p_dz_d);
+            p_dz.add(p_dz_d);
+
+            final Map<String, float[]> p_dwz_d = new HashMap<String, float[]>();
+            p_dwz.add(p_dwz_d);
+
+            for (final String label : _miniBatchDocs.get(d).keySet()) {
+                // init P(z|d,w)
+                float[] p_dwz_dw = ArrayUtils.newRandomFloatArray(_K, _rnd);
+                ArrayUtils.normalize(p_dwz_dw);
```
Comment (myui) [inline core/src/main/java/hivemall/topicmodel/IncrementalPLSAModel.java]: It's better to return a value in `normalize` so that 

```java
import static hivemall.utils.MathUtils.normalize;

float[] p_dwz_dw = normalize(ArrayUtils.newRandomFloatArray(_K, _rnd));
```

can be applied.

File: core/src/main/java/hivemall/utils/math/MathUtils.java [hunk 4]
```diff
@@ -401,4 +401,13 @@ public static double logsumexp(@Nonnull final double[] arr, final double max) {
         return Math.log(logsumexp) + max;
     }
 
+    @Nonnull
+    public static float[] normalize(@Nonnull final float[] arr) {
```
Comment (myui) [inline core/src/main/java/hivemall/utils/math/MathUtils.java]: `sum` is l1norm? Then, remove this to `l1normalize`.

`l1norm` should be the sum of the **absolute** values of the columns.

**Comment by takuti** [comment]
**Observations**:

Current implementation is based on: [Incremental Probabilistic Latent Semantic Analysis for
Automatic Question Recommendation](https://pdfs.semanticscholar.org/b66e/c7faf2e4888503f7ad1537d284f350fb3e58.pdf), and this technique requires us to choose `alpha` as a hyper-parameter used in Eq. (22). However, choosing an appropriate value of `alpha` is difficult for the following reasons:

- In the original formulation Eq. (22), `\sum_w P(w|z)` does not seem to be 1.
- Depending on data and `alpha`, perplexity highly fluctuates (i.e., not monotonically decreasing), and thus checking convergence based on perplexity is difficult.

One possible direction is to consider alternative incremental PLSA techniques. In particular, [Using Incremental PLSI for Threshold-Resilient Online Event Analysis](https://pdfs.semanticscholar.org/a258/b33e285da2e93b59e50311d50ff46045a38b.pdf) is one of the most cited incremental PLSA papers.

I still keep this PR as "WIP" for the reasons that I mentioned above. Will try to implement the alternative technique and report results in a week...

**Comment by takuti** [comment]
I've realized that the main difference between following two papers is in **how to initialize P(w|z) for newly observed words**. 

- [Incremental Probabilistic Latent Semantic Analysis for
Automatic Question Recommendation](https://pdfs.semanticscholar.org/b66e/c7faf2e4888503f7ad1537d284f350fb3e58.pdf) 
- [Using Incremental PLSI for Threshold-Resilient Online Event Analysis](https://pdfs.semanticscholar.org/a258/b33e285da2e93b59e50311d50ff46045a38b.pdf)

The former (i.e., current implementation) simply initializes w/ random values. Previous P(w|z) could be incorporated by setting a hyper-parameter `alpha` if we wanted (, and `alpha=0` is also possible). 

Meanwhile, the latter requires to undergo certain fold-in procedure to compute "better" P(w|z) by setting a window size. IMO, this approach is too much complex to achieve our goal (=implement pLSA UDTF which repeats EM iterations over the same set of mini-batches).

Thus, I will finalize this PR with current implementation.

Todo:

- [ ] Double-check if the algorithm is implemented correctly
- [ ] Documentation
  - Difference with LDA
  - Explain the effect of `alpha`

**Comment by takuti** [comment]
@myui almost finished. could you review current code and docs?

**Comment by myui** [comment]
Sure 👍

**Comment by myui** [comment]
revising at https://github.com/myui/incubator-hivemall/tree/HIVEMALL-74

File: core/src/main/java/hivemall/topicmodel/PLSAPredictUDAF.java [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+            this.wcList = new ArrayList<String>();
+            this.probMap = new HashMap<String, List<Float>>();
+        }
+
+        void iterate(String word, float value, int label, float prob) {
+            wcList.add(word + ":" + value);
+
+            // for an unforeseen word, initialize its probs w/ -1s
+            if (!probMap.containsKey(word)) {
+                List<Float> probEmpty_word = new ArrayList<Float>(Collections.nCopies(topic, -1.f));
```
Comment (myui) [inline core/src/main/java/hivemall/topicmodel/PLSAPredictUDAF.java]: Q. Is this necessary?

```java
List<Float> probEmpty_word = new ArrayList<Float>(Collections.nCopies(topic, -1.f));
```

Why not as follows?

```java
List<Float> probEmpty_word = Collections.nCopies(topic, -1.f);
```

Comment (myui) [inline core/src/main/java/hivemall/topicmodel/PLSAPredictUDAF.java]: ahh... It's immutable.

**Comment by myui** [comment]
@takuti I think probability normalization is missing in `plsa_predict`.

```
[{"label":0,"probability":1.0},{"label":1,"probability":1.1246405E-32}]
```

**Comment by myui** [comment]
@takuti LGTM for other codes. Merged! Thanks.

Please send PR `[HIVEMALL-74-2]` if required.

**Comment by takuti** [comment]
@myui Thanks! Are you planning to push [your revise branch](https://github.com/myui/incubator-hivemall/tree/HIVEMALL-74) to master? Or, can I cherry-pick them and start making additional effort on my own?

For normalization, limitation of floating point numbers?

```java
float[] ary = new float[] {0.99999f, 1e-32f};
System.out.println("Original: " + ary[0] + " " + ary[1]);
double sum = 0.d;
for (int i = 0; i < 2; i++) {
    sum += ary[i];
}
for (int i = 0; i < 2; i++) {
    ary[i] /= sum;
}
System.out.println("Normalized: " + ary[0] + " " + ary[1]);
```

```
Original: 0.99999 1.0E-32
Normalized: 1.0 1.00001E-32
```

**Comment by myui** [comment]
@takuti Revised branch is already merged into master.

**Comment by myui** [comment]
oops... need to fix CI errors as well.
https://travis-ci.org/apache/incubator-hivemall/jobs/226400990

**Comment by takuti** [comment]
Ack! Checking your modifications.

**Comment by takuti** [comment]
^ Fixed and gonna be sent as a new pr soon
